service/history/workflow/update/update.go

```diff
@@ -26,85 +26,392 @@ package update
+		}
+		u.accepted.(*future.FutureImpl[*failurepb.Failure]).Set(rej.Failure, nil)
+		u.outcome.(*future.FutureImpl[*updatepb.Outcome]).Set(&outcome, nil)
+		u.onComplete()
+	})
+	eventStore.OnAfterRollback(func(context.Context) { u.setState(stateRequested) })
+	return nil
+}
+
+// onResponseMsg expectes the Update to be in either stateProvisionallyAccepted
+// or stateAccepted and returns an error if it finds otherwise. On commit of
+// buffered effects the state machine will transtion to stateCompleted and the
+// outcome future is completed with the updatepb.Outcome from the
+// updatepb.Response input message.
+func (u *Update) onResponseMsg(
+	ctx context.Context,
+	res *updatepb.Response,
+	eventStore EventStore,
+) error {
+	if err := u.checkState(res, stateProvisionallyAccepted, stateAccepted); err != nil {
+		return err
+	}
+	if err := validateResponseMsg(u.id, res); err != nil {
+		return err
+	}
+	if _, err := eventStore.AddWorkflowExecutionUpdateCompletedEvent(res); err != nil {
+		return err
+	}
+	u.instrumentation.CountResponseMsg()
+	prevState := u.setState(stateProvisionallyCompleted)
+	eventStore.OnAfterCommit(func(context.Context) {
+		u.setState(stateCompleted)
+		u.outcome.(*future.FutureImpl[*updatepb.Outcome]).Set(res.GetOutcome(), nil)
+		u.onComplete()
+	})
+	eventStore.OnAfterRollback(func(context.Context) { u.setState(prevState) })
+	return nil
+}
+
+func (u *Update) completedOrProvisionallyCompleted() bool {
+	return u.state.IsOneOf(stateCompleted, stateProvisionallyCompleted)
 }
 
-func (u *Update) complete(o *updatepb.Outcome) {
-	u.state = stateCompleted
-	u.pendingOutcome = o
+func (u *Update) hasOutgoingMessage() bool {
+	return u.state.Is(stateRequested)
 }
 
-func (u *Update) reject(f *failurepb.Failure) {
-	u.state = stateRejected
-	u.pendingOutcome = &updatepb.Outcome{
-		Value: &updatepb.Outcome_Failure{
-			Failure: f,
-		},
+func (u *Update) checkState(msg proto.Message, allowedStates ...state) error {
+	if u.state.IsOneOf(allowedStates...) {
+		return nil
 	}
+	return invalidArgf("invalid state transition attempted: "+
+		"received %T message while in state %q", msg, u.state)
+}
+
+// setState assigns the current state to a new value returning the original
+// value.
+func (u *Update) setState(newState state) state {
+	currState := u.state.Load()
+	u.instrumentation.StateChange(u.id, currState, newState)
+	u.state.Set(newState)
+	return currState
+}
+
+func (s state) String() string {
+	switch s {
+	case stateAdmitted:
+		return "Admitted"
+	case stateProvisionallyRequested:
+		return "ProvisionallyRequested"
+	case stateRequested:
+		return "Requested"
+	case stateProvisionallyAccepted:
+		return "ProvisionallyAccepted"
+	case stateAccepted:
+		return "Accepted"
+	case stateProvisionallyCompleted:
+		return "ProvisionallyCompleted"
+	case stateCompleted:
+		return "Completed"
+	}
+	return "unrecognized state"
+}
+
+func (s *state) Is(other state) bool {
```

Comment:
My personal preferences: if type has methods (even trivial) it goes to separate file.

Reply:
I must also say that these 3 methods below are brilliant.

Reply:
Feels like a Java thing to me but I can move it.